java/org/apache/catalina/filters/CsrfPreventionFilter.java

```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+     *        {@link HttpServletResponse#encodeURL(String)}.
+     */
+    public void setNoNonceURLPatterns(String patterns) {
+        this.noNoncePatterns = patterns;
+
+        if (null != context) {
+            this.noNoncePredicates = createNoNoncePredicates(context, this.noNoncePatterns);
+        }
+    }
+
+    /**
+     * Creates a collection of matchers from a comma-separated string of patterns.
+     *
+     * @param patterns A comma-separated string of URL matching patterns.
+     *
+     * @return A collection of predicates representing the URL patterns.
+     */
+    protected static Collection<Predicate<String>> createNoNoncePredicates(ServletContext context, String patterns) {
+        if (null == patterns || 0 == patterns.trim().length()) {
+            return null;
+        }
+
+        String values[] = patterns.split(",");
+
+        ArrayList<Predicate<String>> matchers = new ArrayList<>(values.length);
+        for (String value : values) {
+            Predicate<String> p = createNoNoncePredicate(context, value.trim());
+
+            if (null != p) {
+                matchers.add(p);
+            }
+        }
+
+        matchers.trimToSize();
+
+        return matchers;
+    }
+
+    /**
+     * Creates a predicate that can match the specified type of pattern.
+     *
+     * @param pattern The pattern to match e.g. <code>*.foo</code> or
+     *                <code>/bar/*</code>.
+     *
+     * @return A Predicate which can match the specified pattern, or
+     *         <code>>null</code> if the pattern is null or blank.
+     */
+    protected static Predicate<String> createNoNoncePredicate(ServletContext context, String pattern) {
+        if (null == pattern || 0 == pattern.trim().length()) {
+            return null;
+        }
+        if (pattern.startsWith("mime:")) {
+            return new MimePredicate(context, createNoNoncePredicate(context, pattern.substring(5)));
+        } else if (pattern.startsWith("*")) {
+            return new SuffixPredicate(pattern.substring(1));
+        } else if (pattern.endsWith("*")) {
+            return new PrefixPredicate(pattern.substring(0, pattern.length() - 1));
+        } else if (pattern.startsWith("/") && pattern.endsWith("/")) {
+            return new PatternPredicate(pattern.substring(1, pattern.length() - 1));
+        } else {
+            throw new IllegalArgumentException("Unsupported pattern: " + pattern);
+        }
+    }
+
+    protected static class MimePredicate implements Predicate<String> {
+        private final ServletContext context;
+        private final Predicate<String> predicate;
+
+        public MimePredicate(ServletContext context, Predicate<String> predicate) {
+            this.context = context;
+            this.predicate = predicate;
+        }
+
+        @Override
+        public boolean test(String t) {
+            String mimeType = context.getMimeType(t);
+
+            return predicate.test(mimeType);
```

Comment:
Does predicate handle `null`?

Reply:
Is it often that the application will invoke `HttpServletResponse.encodeURL(null)`?